pallets/asset-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,1010 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use codec::{Encode, Decode, EncodeLike};
+use frame_support::{
+	decl_storage, decl_module,
+	DefaultNoBound,
+	traits::Get,
+	weights::{
+		Weight, DispatchInfo, PostDispatchInfo, GetDispatchInfo, Pays, WeightToFeePolynomial,
+		WeightToFeeCoefficient, DispatchClass,
+	},
+	dispatch::DispatchResult,
+};
+use sp_runtime::{
+	FixedU128, FixedPointNumber, FixedPointOperand, Perquintill, RuntimeDebug,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, ValidTransaction, TransactionValidityError, TransactionValidity,
+	},
+	traits::{
+		Saturating, SignedExtension, SaturatedConversion, Convert, Dispatchable,
+		DispatchInfoOf, PostDispatchInfoOf, Zero, One,
+	},
+};
+use pallet_balances::NegativeImbalance;
+use pallet_transaction_payment::OnChargeTransaction;
+use frame_support::traits::tokens::{fungibles::{Balanced, Inspect, CreditOf}, WithdrawConsequence};
+
+// mod payment;
+// mod types;
+
+// pub use payment::*;
+// pub use types::{InclusionFee, FeeDetails, RuntimeDispatchInfo};
+
+type BalanceOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::Balance;
+type AssetBalanceOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+type AssetIdOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+type LiquidityInfoOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo;
+
+#[derive(Encode, Decode, DefaultNoBound)]
+pub enum InitialPayment<T: Config> {
+	Nothing,
+	Native(LiquidityInfoOf<T>),
+	Asset(CreditOf<T::AccountId, T::Fungibles>),
+}
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		inherent::Vec,
+		traits::{
+			Currency, ReservableCurrency, EnsureOrigin, ExistenceRequirement::KeepAlive,
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_transaction_payment::Config + pallet_balances::Config + pallet_authorship::Config + pallet_assets::Config {
+		type Fungibles: Balanced<Self::AccountId>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> where
+	BalanceOf<T>: FixedPointOperand + Into<AssetBalanceOf<T>>,
+	AssetBalanceOf<T>: FixedPointOperand,
+{
+	fn to_asset_balance(balance: BalanceOf<T>, asset_id: AssetIdOf<T>) -> Result<AssetBalanceOf<T>, ()> {
+		let res = pallet_assets::BalanceToAssetBalance::<T, BalanceOf<T>, <T as pallet_balances::Config>::ExistentialDeposit>::to_asset_balance(balance, asset_id)
```

Comment:
Bird's eye view, I'd add `type BalanceToAsset: Convert<Balance, Asset>` (or a custom trait) to the config of `asset-tx-payment` rather than relying on `pallet-assets` directly.

Reply:
Can you elaborate what this pallet will do? 

I know that the goal is to pay tx-fees in an asset, but why is there a conversion of a balance to asset? if anythign I'd expect to see the opposite: we need to convert the asset to balance, and then relay it to the normal tx-payment that uses a balance, or somethign along those lines.

Reply:
No, you have it backwards.
We already have fee calculation logic that determines the fee in terms of the native currency. What we now want to allow is paying tx fees with an asset, so what we need to do is convert the fee from native to asset and then charge the account sending the transaction.
Most of that should already be implemented in this pallet.

Reply:
Can't we tweak the current `pallet-transaction-payment` to: 

instead of `type Currency: Currency<_>`, work with a more generalized trait that is implemented by both `pallet-assets` and `pallet-balances`. Does such a thing exist?

Reply:
But we need custom data in the transaction: the asset id